File: ios/opensleep/opensleep/Services/Sleep/Percentile.swift
import Foundation

/// Port of com.urbandroid.util.Percentile from Sleep as Android.
/// Evaluates the p-th percentile of a collection or array of floats with linear interpolation.
final class Percentile {
    private var quantile: Float

    init(quantile: Float = 50.0) {
        self.quantile = quantile
        setQuantile(quantile)
    }

    func setQuantile(_ f: Float) {
        precondition(f >= 0.0 && f <= 100.0, "Illegal quantile value: \(f)")
        self.quantile = f
    }

    private func evaluateSorted(_ arr: [Float], _ f: Float) -> Float {
        let length = Float(arr.count)
        let f2 = ((length + 1.0) * f) / 100.0
        let fFloor = floor(f2)
        let i = Int(fFloor)
        let f3 = f2 - fFloor

        if f2 < 1.0 {
            return arr[0]
        }
        if f2 >= length {
            return arr[arr.count - 1]
        }
        let f4 = arr[i - 1]
        return ((arr[i] - f4) * f3) + f4
    }

    func evaluate(_ list: [Float], _ i: Int) -> Float {
        if i <= 0 || i > 100 { return 0.0 }
        if list.isEmpty { return .nan }
        if list.count == 1 { return list[0] }

        let sorted = list.sorted()
        return evaluateSorted(sorted, Float(i))
    }

    func evaluate(_ arr: [Float], from: Int, to: Int, _ f: Float) -> Float {
        let len = to - from
        if f <= 0.0 || f > 100.0 { return 0.0 }
        if len == 0 { return .nan }
        if len == 1 { return arr[from] }

        let slice = Array(arr[from..<to]).sorted()
        return evaluateSorted(slice, f)
    }

    func evaluate(_ arr: [Float], _ f: Float) -> Float {
        evaluate(arr, from: 0, to: arr.count, f)
    }
}

File: android/app/src/main/java/app/opensleep/domain/sleep/Percentile.kt
package app.opensleep.domain.sleep

import java.util.Arrays

/**
 * Port of com.urbandroid.util.Percentile from Sleep as Android.
 * Evaluates the p-th percentile of a collection or array of floats with linear interpolation.
 */
class Percentile(private var quantile: Float = 50.0f) {

    init {
        setQuantile(quantile)
    }

    fun setQuantile(f: Float) {
        require(f in 0.0f..100.0f) { "Illegal quantile value: $f" }
        this.quantile = f
    }

    private fun evaluateSorted(fArr: FloatArray, f: Float): Float {
        val length = fArr.size
        val f2 = ((length + 1.0f) * f) / 100.0f
        val fFloor = Math.floor(f2.toDouble()).toFloat()
        val i = fFloor.toInt()
        val f3 = f2 - fFloor
        if (f2 < 1.0f) {
            return fArr[0]
        }
        if (f2 >= length) {
            return fArr[length - 1]
        }
        val f4 = fArr[i - 1]
        return ((fArr[i] - f4) * f3) + f4
    }

    fun evaluate(list: List<Float>, i: Int): Float {
        if (i <= 0 || i > 100) return 0.0f
        if (list.isEmpty()) return Float.NaN
        if (list.size == 1) return list[0]

        val fArr = FloatArray(list.size)
        for (idx in list.indices) {
            fArr[idx] = list[idx]
        }
        Arrays.sort(fArr)
        return evaluateSorted(fArr, i.toFloat())
    }

    fun evaluate(fArr: FloatArray, fromIndex: Int, toIndex: Int, f: Float): Float {
        val len = toIndex - fromIndex
        if (f <= 0.0f || f > 100.0f) return 0.0f
        if (len == 0) return Float.NaN
        if (len == 1) return fArr[fromIndex]

        val copy = FloatArray(len)
        System.arraycopy(fArr, fromIndex, copy, 0, len)
        Arrays.sort(copy)
        return evaluateSorted(copy, f)
    }

    fun evaluate(fArr: FloatArray, f: Float): Float {
        return evaluate(fArr, 0, fArr.size, f)
    }
}
